### PR TITLE
easy and fast way to handle dups in list

### PR DESCRIPTION
**Pull request by UnknownCrafts**
Way faster than running a for loop while checking a dictionary to make a duplicate free list, my method also preserves the order of the list.

### CODE REVIEW

File: pytube/streams.py [hunk 1]
```diff
@@ -271,6 +271,7 @@ def download(
                     bytes_remaining -= len(chunk)
                     # send to the on_progress callback.
                     self.on_progress(chunk, fh, bytes_remaining)
+            fh.close()
```
Comment (glubsy) [inline pytube/streams.py]: You most likely do not need to call close() on the file handle if you already use a context manager (the `open(...) as fh` above). The context manager calls it automatically for you.

File: pytube/helpers.py [hunk 2]
```diff
@@ -331,5 +325,6 @@ def create_mock_html_json(vid_id) -> Dict[str, Any]:
     logger.info(f'Outputing json.gz file to {gzip_filepath}')
     with gzip.open(gzip_filepath, 'wb') as f:
         f.write(json.dumps(html_data).encode('utf-8'))
+        f.close()
```
Comment (glubsy) [inline pytube/helpers.py]: You most likely do not need to call close() on the file handle if you already use a context manager (the open(...) as f above). The context manager calls it automatically for you.

Comment (UnknownCrafts) [reply pytube/helpers.py]: yea i just realized that "with" already does that. mb

File: pytube/contrib/search.py [hunk 3]
```diff
@@ -39,10 +39,9 @@ def completion_suggestions(self):
         :returns:
             A list of autocomplete suggestions provided by YouTube for the query.
         """
-        if self._completion_suggestions:
-            return self._completion_suggestions
```
Comment (glubsy) [inline pytube/contrib/search.py]: This does not seem to be a redundant statement. It avoids updating the cache if the value has been set once.

Comment (UnknownCrafts) [reply pytube/contrib/search.py]: oh

**Comment by UnknownCrafts** [comment]
What do you think about the "easy and fast way to handle dups" commit

**Comment by glubsy** [comment]
> What do you think about the "easy and fast way to handle dups" commit

I see 2 approaches to this question: 1) performance 2) readability

1) I have not benchmarked this, but I doubt this adds much in terms of performance

2) Sure it shortens the function, and became more "readable" in a way, but it also became less intelligible. Someone with less python experience would have to lookup what fromkeys() does and why the OrderedDict is used here.

I think it's alright.

**Comment by UnknownCrafts** [comment]
by "string.count" i meant "list.count"

**Comment by UnknownCrafts** [comment]
btw there is a weird error for me in extract.py
Screen Shot 2022-01-31 at 8 57 31 AM

**Comment by UnknownCrafts** [comment]
not sure if you have it too. I haven't even touched that line

**Comment by UnknownCrafts** [comment]
> > What do you think about the "easy and fast way to handle dups" commit
> 
> I see 2 approaches to this question: 1) performance 2) readability
> 
> 1. I have not benchmarked this, but I doubt this adds much in terms of performance
> 2. Sure it shortens the function, and became more "readable" in a way, but it also became less intelligible. Someone with less python experience would have to lookup what fromkeys() does and why the OrderedDict is used here.
> 
> I think it's alright.

I benchmarked it and found out that the performance difference grows depending on how many times the functions are iterated and not the size of the list.
Screen Shot 2022-01-31 at 9 03 47 AM

**Comment by UnknownCrafts** [comment]
it was a quick benchmark so there might be some other aspects that can still be checked.

**Comment by glubsy** [comment]
Regarding https://github.com/pytube/pytube/pull/1215/commits/fed790557299513d3c27c7f8b662c7c387559bfb, same thing as above, it became a bit less readable, but it's alright in my opinion.

Thanks for taking the time to benchmark your changes.

A few more remarks:

* Better keep separate changes as separate pull requests if possible.
* Post your test code as a gist so that everyone can test on their machine as well. However, make sure to choose more descriptive symbol names. It is a bit difficult to make sense of the variable names at first.
* Regarding the Pylance warning (I don't use Pylance) it's not that important but it is indeed true that it should be Tuple(str, str) instead of (str, str).

**Comment by UnknownCrafts** [comment]
alright i will open other pull requests if I change other stuff

File: pytube/extract.py [hunk 4]
```diff
@@ -67,10 +64,8 @@ def is_private(watch_html):
         "\"simpleText\":\"Private video\"",
         "This video is private."
     ]
-    for string in private_strings:
-        if string in watch_html:
-            return True
-    return False
+
```
Comment (glubsy) [inline pytube/extract.py]: Just nitpicking, but  it's best to keep the same consistent line spacing style, so I would remove that empty line.

Comment (UnknownCrafts) [reply pytube/extract.py]: ight

**Comment by UnknownCrafts** [comment]
that would be the last commit for this branch, I will make another branch now

**Comment by UnknownCrafts** [comment]
`from collections import OrderedDict
import time
from typing import Any, Dict


#functions

def useDict(dup_list): 
    return list(OrderedDict.fromkeys(dup_list))

def forLoop(dup_list):
    seen: Dict[Any, bool] = {}
    result = []
    for item in dup_list:
        if item in seen:
            continue
        seen[item] = True
        result.append(item)
    return result

#Simple benchmark
iterations = 1000
_list = [1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1]

#benchmark useDict()
start1 = time.time()
for i in range(iterations): useDict(_list)
end1 = time.time()

#benchmark forLoop()
start2 = time.time()
for i in range(iterations): forLoop(_list)
end2 = time.time()

#get the total time spent in the functions
total_time_useDict = start1 - end1
total_time_forLoop = start2 - end2

#print the result
print("useDict: " + str(total_time_useDict) + " vs " + "forLoop: " + str(total_time_forLoop), "iterations: " + str(iterations))`

**Comment by UnknownCrafts** [comment]
`
from collections import OrderedDict
import time
from typing import Any, Dict


#functions

def useDict(dup_list): 
    return list(OrderedDict.fromkeys(dup_list))

def forLoop(dup_list):
    seen: Dict[Any, bool] = {}
    result = []
    for item in dup_list:
        if item in seen:
            continue
        seen[item] = True
        result.append(item)
    return result

#Simple benchmark
iterations = 1000
_list = [1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1,1, 5, 3, 6, 3, 5, 6, 1]

#benchmark useDict()
start1 = time.time()
for i in range(iterations): useDict(_list)
end1 = time.time()

#benchmark forLoop()
start2 = time.time()
for i in range(iterations): forLoop(_list)
end2 = time.time()

#get the total time spent in the functions
total_time_useDict = start1 - end1
total_time_forLoop = start2 - end2

#print the result
print("useDict: " + str(total_time_useDict) + " vs " + "forLoop: " + str(total_time_forLoop), "iterations: " + str(iterations))
`

**Comment by UnknownCrafts** [comment]
that is the test code i used

**Review by glubsy — APPROVED** [review]
Looks good.

**Review by UnknownCrafts — COMMENTED** [review]
alright

**Comment by UnknownCrafts** [comment]
is someone gonna approve the workflow? I wanna delete this branch soon

**Comment by RONNCC** [comment]
hey @UnknownCrafts 


Thanks for the PR!

If this is still relevant, could you add the following:

- a unit test to show the behavior before and after
- some context on the need for the change

Ideally this would 'break' in the future when youtube also changes behavior again and we can show it as a regression.
In addition it would help document the change in the code using tests.

**Pull request closed without merging**